Add pre-commit hooks and enforce code formatting

mbollmann opened the pull request:
**This PR is a suggestion – please discuss!**

With this PR, developers can install a bunch of useful git pre-commit hooks by simply running `make install_hooks` on their local machine. These hooks are installed and managed via [the pre-commit tool](https://pre-commit.com/).

Right now, it **adds the following hooks:**

+ `black`, to run the Python code formatter [Black](https://black.readthedocs.io/en/stable/).

   Enforcing a uniform coding style is a good thing, especially with a code base as large as this one, and I happen to quite like Black – although it's of course not the only option. I'm happy to hear your thoughts.

+ `check-yaml` and `check-xml`, to perform basic syntax checks for our YAML and XML files.

   This has already uncovered that one of our YAML files had duplicate entries, which happened to be harmless, but still shouldn't occur.

+ `end-of-file-fixer` and `trailing-whitespace`, to make sure files end in exactly one newline and lines have no trailing whitespace.

   I think this is generally a good idea. I don't know if I'm the only one here, but I have my editor set up to do this automatically when saving any file, potentially introducing noise into future commits if files didn't adhere to this to begin with. (That doesn't mean I want to force my preferences on everyone, but as I said, I think it's a good idea anyway.)

+ `requirements-txt-fixer`, a tiny hook that makes sure requirements are alphabetically ordered.

+ `no-commit-to-branch`, which is set up to prevent accidental commits to master.

You can see in the respective commits on this PR which files would currently be changed by these hooks. If people approve of merging this (or a variant of this), I would also add documentation to the respective READMEs.

**Potential downsides:**

+ Merging this would introduce a lot of noise in the commit history as a one-off event, though hopefully make for a cleaner repo in the future.

+ Third-party scripts that we added to our repo are also affected. If we don't like this, we could move them to a separate subdirectory which would make them easy to exclude with a regex pattern.

akoehn commented:
Generally a good idea! I can already say that I will have comments tomorrow.

akoehn commented:
My thoughts on this: pre-commit-hooks should check but not change.

I find it acceptable to e.g. have a make target "fix-code-style" or the like and make the pre-commit-hook only check whether everything is fine (but I would prefer the check to be a Makefile target as well so that it can be invoked as needed easily). This would also make it easy to run the checks on PRs.

The kind of fixes all seem fine to me in principle, but some fine-tuning especially regarding black might be beneficial.

My proposal: Agree on whether we want pre-commit checks (I am for it), pre-commit fixes (I am against it), and then discuss the checks and fixes afterwards.

mbollmann commented:
The point of this would be to have cleaner commits and a more consistent code style, so presumably we would also run these checks with Travis and fail PRs if they don't comply. In this case, having to remember to  run the fixes manually each time seems like a hassle and likely source of frustration. The whole point of the pre-commit setup is to make this effortless.

Also note that no one can be forced to install the hooks. If you prefer to run them manually, you can still do that - just don't use the hooks.

akoehn commented:
As I said: I am fine with running *checks* in pre-commit hooks, but the pre-commit hook should not *change* files.

mbollmann commented:
I understood and respect that, but what is the reasoning?

akoehn commented:
The reasoning is that a system should not do something unexpected.  When I perform a commit, I expect a commit to happen (or fail), not that files change.  

Also, in some cases I might be unhappy with the decisions taken by the auto-correction  but I don't see them and therefore cannot easily intervene.

My proposal: Have a make target, which is also run by a pre-commit hook.  If everything is fine, it just prints "yay, your code looks good!", otherwise it shows the specific complains and appends "You can automatically fix these problems by running `make autofix` and inspect the changes."

This means:
 - code can be checked easily during development without making a commit
 - only good code is commited
 - fixing code is easy if desired
 - code is only changes when the user actually requests it

akoehn commented:
Also, when creating make targets, every user can decide whether they want to add the `check` or the `autofix` target as pre-commit hook and we don't need this discussion :-)

mbollmann commented:
> Also, in some cases I might be unhappy with the decisions taken by the auto-correction but I don't see them and therefore cannot easily intervene.

To be clear: commits always fail when a pre-commit hooks changed files, and their output shows at a minimum the files that they changed. You need to explicitly approve by re-committing.

> My proposal: Have a make target, which is also run by a pre-commit hook. If everything is fine, it just prints "yay, your code looks good!", otherwise it shows the specific complains and appends "You can automatically fix these problems by running `make autofix` and inspect the changes."

If our Travis setup will *require* that the code passes these checks, I don't know why I would ever want to commit something without running the fixes. *Having* to run this manually seems like an annoying and unnecessary burden to me, but I guess if we haven't convinced each other by now, we won't, so maybe others should weigh in.

akoehn commented:
> I don't know why I would ever want to commit something without running the fixes.

I never suggested that. In my opinion the user should have the freedom to decide whether the fixes should be done *automatically* or *manually* on violation or not. That they do need fixing is not under discussion.

Consider this made-up example: User A edits a .yaml file and mis-types by accident, writing `acl2020` instead of `eacl2020`, but `acl2020` already exists.  The autofix will automatically remove the second `acl2020` and the user might be confused.

As the checks need to be run by CI anyway, they should be make targets (for consistency).  I currently fail to see the downside of creating a fixing and a checking make target and then letting the user decide whether they want to enable the fixing or the checking pre-commit hook.

mbollmann commented:
> Consider this made-up example: User A edits a .yaml file and mis-types by accident, writing `acl2020` instead of `eacl2020`, but `acl2020` already exists. The autofix will automatically remove the second `acl2020` and the user might be confused.

No, it won't. I believe none of the hooks will perform potentially destructive changes (and if so, I'd recommend not to use them).

> I currently fail to see the downside of creating a fixing and a checking make target and then letting the user decide whether they want to enable the fixing or the checking pre-commit hook.

That sounds like a workable solution indeed, although we'd have to find an alternative to the pre-commit tool, since most of its hooks don't support checking without fixing.

mbollmann commented:
Sorry, that last part may have been confusing. Pre-commit hooks named `check-*` will only check, never fix by themselves; black can certainly be set to only check; but the other convenience hooks like `trailing-whitespace` etc. do not have a "check-only" mode.

akoehn commented:
> I believe none of the hooks will perform potentially destructive changes (and if so, I'd recommend not to use them).

Okay, I went by the changes in the PR and assumed they were automatic.  I am perfectly fine with automatic trailing-whitespace and newline fixing as well as requirements-txt-fixer (I don't see the point of that one but I also don't care), it is only the ones that change more than that I have a problem with (namely black as the others as pointed out only check).

mjpost commented on bin/add_attachment.py:
```diff
@@ -40,86 +40,118 @@
+ALLOWED_TYPES = ["pdf", "pptx", "zip"]
+ATTACHMENT_TYPES = "Poster Presentation Note Software Supplementary Dataset".split()
+
 
```
[PEP8 doesn't specify](https://www.python.org/dev/peps/pep-0008/#id25) whether to use single or double quotes. We could save a lot of commit history by allowing either.

mjpost commented on bin/add_attachment.py:
```diff
@@ -40,86 +40,118 @@
+                "-> FATAL: An SSL error was encountered in downloading the files.",
+                file=sys.stderr,
+            )
             sys.exit(1)
```
What is the rule that triggers the print formatting here? If it is line length, I personally would much prefer

```python
print("-> FATAL: An SSL error was encountered in downloading the files.",
      file=sys.stderr)
```

But I can be convinced to update my personal preferences if there is a good rule.

mjpost commented:
+1 on everything here, except black.

That's not to say "no" to black, but it may take some adjusting. As I noted above, most of the commit changes are a result of black changing quotes, which PEP8 has nothing to say about (though I find [black's reasoning](https://black.readthedocs.io/en/stable/the_black_code_style.html#strings) convincing enough). I agree that coding conventions are good, and would for example like to enforce XML indentation, spaces instead of tabs (even though tabs are more developer-friendly), but the quotes and the `print()` statement formatting bug me at the moment.

Anyway, if it were me deciding, I'd take everything but black. Sometimes there's an argument for allowing people a little freedom.

mbollmann commented:
Thanks for the comments, @mjpost.

Black is, by design, a very opinionated code formatter, so there are almost no configuration options. Incidentally, the two things that can be configured are the ones you mention. :)

+ Regarding **double quotes vs. single quotes:** I don't like the visual inconsistency from having a mix of both of these without a clear reason ([example](https://github.com/acl-org/acl-anthology/blob/7f96661ef740e52e8145eebe235faffaa4e6586b/bin/add_dois.py#L133)). It's not a hill I want to die on though, and this is one behavior of black that can be turned off.

+ The print statement formatting is indeed just **line length.** [The behavior of the line wrapping component is explained here.](https://github.com/psf/black#how-black-wraps-lines) As for why it wraps the way it does:

  > You might have noticed that closing brackets are always dedented and that a trailing comma is always added. Such formatting produces smaller diffs; when you add or remove an element, it's always just one line. Also, having the closing bracket dedented provides a clear delimiter between two distinct sections of the code that otherwise share the same indentation level

  What can be configured is the number of characters that trigger wrapping; the default setting is 88.

If we don't like the behavior of black outside of these two configurable options, we'd need to look into a different code formatter (such as Google's YAPF).

mjpost commented:
I agree about quotes. I'd been using single-quotes by default (that's what I used when working on Sockeye at Amazon), but have been moving to double-quotes since that seems to be your and others' convention here. I'd rather nudge people in that direction that enforce it, if we can agree on that. The thing I really don't want is for black to prevent a commit because I've chosen to use quotes it doesn't like (say, to avoid backslashes). I know the docs say they explicitly include this in their calculation, so maybe there is nothing to worry about.

As for line length, I could tolerate the formatting...but maybe in acknowledgement of our history we could set it to 110?

These aren't hills I'm willing to die on, either, though. The bottom line is that I have ambivalent preferences here, but think that code standards are good and am willing to adjust if we have consensus.

mbollmann commented:
With the recent commit, I've added the full pre-commit check suite to the `check` target, with the change that black runs in check-only mode now. Checks are run over *all* files in the repo, so this is suitable for checking with Travis.

I've also introduced a `check_commit` target that only checks files currently staged for committing, which is considerably faster and should probably become our default pre-commit hook. I'm not sure if this is the best way of making this distinction in the Makefile, but I think it's a useful one to make.

I haven't introduced an `autofix` target yet, as I don't see a way to run hooks with different arguments via pre-commit (in this case: `black` with vs. without `--check`); I'd probably suggest un-coupling the code formatter/checker from the pre-commit suite and running it separately. But for that, we should decide on a formatter first.

(I'm not wedded to black and happy to switch to something else if @akoehn or others disagree with its philosophy/lack of configurability.)

mbollmann commented:
> I agree about quotes. I'd been using single-quotes by default (that's what I used when working on Sockeye at Amazon), but have been moving to double-quotes since that seems to be your and others' convention here.

I must admit I find it difficult to follow a convention, which is why I like to have a code formatter that takes care of this. ;)

> The thing I really don't want is for black to prevent a commit because I've chosen to use quotes it doesn't like (say, to avoid backslashes).

Not to beat a dead horse, but that's why I prefer autofixing hooks, since it's a non-issue that way.

akoehn commented:
>>The thing I really don't want is for black to prevent a commit because I've chosen to use quotes it doesn't like (say, to avoid backslashes).
>
>Not to beat a dead horse, but that's why I prefer autofixing hooks, since it's a non-issue that way.

I think @mjpost does not want to be overruled, be it manually or by autofixing; I did not have time to look into the rules black enforces but had a look at in the summer and can't remember negative emotions, so feel free to make it enforce its rules; I won't complain.

But please do not make the line length 110, that seems to be decidedly too much when working with more than one column of data and it encourages complex single-line constructs that are hard to read afterwards. Am happy to be overruled by that as well, but I would pick 90.

akoehn commented:
I had a further look at the style black imposes and find it reasonable enough to work with.  It is certainly better than no consistent style and formatting rules are not important for me to think about for a longer period of time.  That is enough for me to say yes to black.

mjpost commented:
I'm increasingly coming around to all of this, and in any case trust your (pl.) judgment. I'm fine to merge and will defer to you.

mbollmann commented:
Changing the line length from 88 to 90 already removes a lot of the newly introduced line breaks. I'm happy to go with that.

Before I merge, I'd like to:

+ Add the `skip-string-normalization` flag – @mjpost, should I also revert all the string quote normalizations already performed in this PR, or do you think it's okay to normalize them all here but just not enforce it in the future?

+ Add the `autofix` target

+ Add a git template that has `make check_commit` as a pre-commit hook for all newly-cloned repos

+ Add some documentation to READMEs about this

This should all not take a lot of work.

akoehn commented:
Am 28. November 2019 22:25:22 MEZ schrieb Marcel Bollmann <notifications@github.com>:
>Changing the line length from 88 to 90 already removes a lot of the
>newly introduced line breaks. I'm happy to go with that.>
>>
>Before I merge, I'd like to:>
>>
>+ Add the `skip-string-normalization` flag – @mjpost, should I also
>revert all the string quote normalizations already performed in this
>PR, or do you think it's okay to normalize them all here but just not
>enforce it in the future?

I think we can also enforce that in the future. The case described by @mjpost (writing 'foo"bar' to not have to escape the better quote) is accepted by black.

mbollmann commented:
I have two Makefile/shell questions @akoehn, maybe you can help me out:

+ On `make check_commit`, I'd like to run black only on Python files with staged changes. I don't think I can do that with a single `git diff` command, and would normally do `git diff --staged --name-only bin/ | egrep "\.py$"`, but it seems pipes don't work within `$(shell ... )` constructs?

+ On `make autofix`, I'd like to *always* run two commands (`pre-commit` and `black`) no matter their exit codes, but have the target fail with a non-zero exit code when *either* of them fails. I can't seem to figure out a non-ugly way to do that.

Any ideas would be appreciated!

akoehn commented:
Marcel Bollmann writes:

> +  `git diff --staged --name-only bin/ | egrep "\.py$"`, but it seems
> pipes don't work within `$(shell ... )` constructs?

They should work just fine, but you have an unescaped $ sign in there. $
denotes the start of a Makefile variable and if your command should have
a $, you need to write $$ instead.

> + On `make autofix`, I'd like to *always* run two commands
> (`pre-commit` and `black`) no matter their exit codes, but have the
> target fail with a non-zero exit code when *either* of them fails. I
> can't seem to figure out a non-ugly way to do that.

My suggestion would be something like this:
https://stackoverflow.com/questions/16080716/execute-multiple-commands-in-a-bash-script-sequentially-and-fail-if-at-least-one#16081218

mbollmann commented:
Thanks so much @akoehn!

Of course it wasn't as easy as I thought — turns out black can only return a non-zero exit status with `--check`, not when actually modifying. So I dug into the pre-commit tool and found out how it checks whether a hook modified files (= via comparing `git diff` outputs) and replicated that. Not the most elegant procedure, I guess, but I don't see a nicer way if we want to have separate `check_commit` and `autofix` targets.

If @akoehn can live with this Makefile configuration, I'll add some documentation for this and merge tomorrow. :)

mbollmann commented:
Ah, good point, I can see how that could be annoying while developing. pre-commit seems to do `git ls-files -z` to get its list of files.

I'll also have to investigate why the Travis PR build seems to get stuck in this branch...
